server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,214 @@
+        return masterHistoryCopy.isEmpty() ? null : masterHistoryCopy.get(masterHistoryCopy.size() - 1).master;
+    }
+
+    /**
+     * Returns the most recent non-null master seen, or null if there has been no master seen. Only 30 minutes of history is kept. If the
+     * most recent master change is more than 30 minutes old and that change was to set the master to null, then null will be returned.
+     * @return The most recent non-null master seen, or null if there has been no master seen.
+     */
+    public @Nullable DiscoveryNode getMostRecentNonNullMaster() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        DiscoveryNode mostRecentNonNullMaster = null;
+        for (int i = masterHistoryCopy.size() - 1; i >= 0; i--) {
+            TimeAndMaster timeAndMaster = masterHistoryCopy.get(i);
+            if (timeAndMaster.master != null) {
+                mostRecentNonNullMaster = timeAndMaster.master;
+                break;
+            }
+        }
+        return mostRecentNonNullMaster;
+    }
+
+    /**
+     * Returns true if for the life of this MasterHistory (30 minutes) only one non-null node has been master, and the master has switched
+     * from that node to null n times.
+     * @param n The number of times the non-null master must have switched to null
+     * @return True if there has been a single non-null master and it has switched to null n or more times.
+     */
+    public boolean hasSameMasterGoneNullNTimes(int n) {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        return hasSameMasterGoneNullNTimes(
+            masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).collect(Collectors.toList()),
+            n
+        );
+    }
+
+    public static boolean hasSameMasterGoneNullNTimes(List<DiscoveryNode> masters, int n) {
+        if (getDistinctMastersSeen(masters).size() != 1) {
+            return false;
+        }
+        boolean seenNonNull = false;
+        int timesMasterHasGoneNull = 0;
+        for (DiscoveryNode master : masters) {
+            if (master != null) {
+                seenNonNull = true;
+            } else if (seenNonNull) {
+                timesMasterHasGoneNull++;
+            }
+        }
+        return timesMasterHasGoneNull >= n;
+    }
+
+    private static Set<DiscoveryNode> getDistinctMastersSeen(List<DiscoveryNode> masters) {
+        return masters.stream().filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns the set of distinct non-null master nodes seen in this history.
+     * @return The set of all non-null master nodes seen. Could be empty
+     */
+    public Set<DiscoveryNode> getDistinctMastersSeen() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        return getDistinctMastersSeen(masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).collect(Collectors.toList()));
```

Comment:
Could we not filter here and return the Set ? (it'd be rather inefficient to have this create an intermediate list and then recreate the stream and the final set in `getDistinctMastersSeen`)

Reply:
I don't think we're talking about very performance-sensitive code here (it ought to be a pretty small list that's accessed rarely) so I hadn't focused on that. But I can make this a little more optimized.